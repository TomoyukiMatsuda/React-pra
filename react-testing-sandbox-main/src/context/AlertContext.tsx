import { createContext, useCallback, useContext, useReducer } from "react"

type AlertState = {
  show: boolean
  message: string
}
type AlertDispatch = {
  showDispatcher: (message: string) => void
  hideDispatcher: () => void
}
const initialState = {
  show: false,
  message: ""
}
const alertStateContext = createContext<AlertState>(initialState)
const alertDispatchContext = createContext<AlertDispatch>({
  showDispatcher: () => void 0,
  hideDispatcher: () => void 0
})

export const useAlertState = () => useContext(alertStateContext)
export const useAlertDispatch = () => useContext(alertDispatchContext)

export function AlertProvider({ children }: { children: JSX.Element[] | JSX.Element }) {
  const [state, dispatch] = useReducer(reducer, initialState)
  const showDispatcher = useCallback((message: string) => dispatch(showAlert(message)), [])
  const hideDispatcher = useCallback(() => dispatch(hideAlert()), [])
  return (
    <alertStateContext.Provider value={state}>
      <alertDispatchContext.Provider value={{ showDispatcher, hideDispatcher }}>
        {children}
      </alertDispatchContext.Provider>
    </alertStateContext.Provider>
  )
}

type AlertAction = ReturnType<typeof showAlert> | ReturnType<typeof hideAlert>

function reducer(state: AlertState, action: AlertAction) {
  switch (action.type) {
    case "ui/alert:show":
      return {
        ...state,
        show: true,
        message: action.payload.message
      }
    case "ui/alert:hide":
      return {
        ...state,
        show: false
      }
    default:
      return state
  }
}

function showAlert(message: string) {
  return {
    type: "ui/alert:show",
    payload: {
      message
    }
  } as const
}

function hideAlert() {
  return {
    type: "ui/alert:hide"
  } as const
}
