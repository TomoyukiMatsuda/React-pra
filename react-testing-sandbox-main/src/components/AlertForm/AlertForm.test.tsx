import { cleanup, fireEvent, render } from "@testing-library/react"

import * as AlertContext from "~/context/AlertContext"

import { AlertForm } from "."

let useAlertDispatchSpy: jest.SpyInstance<unknown>
let alertShowDispatchSpy: jest.SpyInstance<unknown>
let alertHideDispatchSpy: jest.SpyInstance<unknown>

describe("AlertForm", () => {
  beforeEach(() => {
    useAlertDispatchSpy = jest.spyOn(AlertContext, "useAlertDispatch")
    alertShowDispatchSpy = jest.fn()
    alertHideDispatchSpy = jest.fn()
    useAlertDispatchSpy.mockImplementation(() => ({
      showDispatcher: alertShowDispatchSpy,
      hideDispatcher: alertHideDispatchSpy
    }))
  })
  afterEach(() => {
    useAlertDispatchSpy.mockClear()
    alertShowDispatchSpy.mockClear()
    alertHideDispatchSpy.mockClear()
    cleanup()
  })
  test("render", () => {
    const { asFragment } = render(<AlertForm />)
    expect(asFragment()).toMatchSnapshot()
  })
  test("click: show dispatch", () => {
    const { container } = render(<AlertForm />)
    const input = container.querySelector("input") as HTMLInputElement
    const button = container.querySelector("button") as HTMLButtonElement
    fireEvent.change(input, { target: { value: "test message" } })
    fireEvent.click(button)
    expect(alertShowDispatchSpy).toBeCalledWith("test message")
  })
  test("submit: show dispatch", () => {
    const { container } = render(<AlertForm />)
    const input = container.querySelector("input") as HTMLInputElement
    const form = container.querySelector("form") as HTMLFormElement
    fireEvent.change(input, { target: { value: "test message" } })
    fireEvent.submit(form)
    expect(alertShowDispatchSpy).toBeCalledWith("test message")
  })
  test("click: hide dispatch", () => {
    const { getByText } = render(<AlertForm />)
    const button = getByText("Alert Close")
    fireEvent.click(button)
    expect(alertHideDispatchSpy).toBeCalled()
  })
})
