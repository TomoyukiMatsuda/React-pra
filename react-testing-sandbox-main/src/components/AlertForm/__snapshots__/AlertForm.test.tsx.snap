// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertForm render 1`] = `
<DocumentFragment>
  <form>
    <input
      type="text"
      value=""
    />
    <button>
      Alert
    </button>
    <br />
    <button
      type="reset"
    >
      Alert Close
    </button>
  </form>
</DocumentFragment>
`;
